### PR TITLE
eleminate --user-compset, allow longname in tests

### PR DESCRIPTION
**Pull request by jedwards4b**
Remove the --user-compset arg, it's no longer needed.   Allow create_test to accept compset longname as part of the testname, add test to scripts_regression_tests.py of these features.
I also change some more internal cime calls to xmlquery to use --value instead of -value.

Test suite: scripts_regression_tests.py
Test baseline: 
Test namelist changes: 
Test status: bit for bit
Fixes ESMCI/cime#1718 
Fixes ESMCI/cime#1706 

User interface changes?: Remove --user-compset from create_newcase, allow compset longname in testname argument to create_test 

Update gh-pages html (Y/N)?: N

Code review:

### CODE REVIEW

File: config/cesm/machines/Makefile [hunk 1]
```diff
@@ -852,7 +852,7 @@ realclean: clean cleancsmshare cleanpio cleanmct cleangptl
 ifneq ($(MAKECMDGOALS), db_files)
 ifneq ($(MAKECMDGOALS), db_flags)
 ifeq (,$(findstring clean,$(MAKECMDGOALS)))
-    -include $(CURDIR)/Depends $(CASEROOT)/Depends.$(COMPILER) $(CASEROOT)/Depends.$(MACH)
+    -include Depends $(CASEROOT)/Depends.$(COMPILER) $(CASEROOT)/Depends.$(MACH)
 endif
```
Comment (gold2718) [inline config/cesm/machines/Makefile]: Are the changes to Makefile required for this PR or they unrelated?

File: src/build_scripts/buildlib.csm_share [hunk 2]
```diff
@@ -13,22 +13,22 @@ my ($sharedlibroot, $installroot, $CASEROOT) = @ARGV;
...
+my $NINST_ICE		= `./xmlquery NINST_ICE		--value`;
+my $NINST_GLC		= `./xmlquery NINST_GLC		--value`;
+my $NINST_LND		= `./xmlquery NINST_LND		--value`;
+my $NINST_OCN		= `./xmlquery NINST_OCN		--value`;
+my $NINST_ROF		= `./xmlquery NINST_ROF		--value`;
+my $NINST_WAV		= `./xmlquery NINST_WAV		--value`;
+my $NINST_ESP		= `./xmlquery NINST_ESP		--value`;
+my $NINST_VALUE		= `./xmlquery NINST_VALUE	--value`;
+$ENV{PIO_VERSION}  	= `./xmlquery PIO_VERSION	--value`;
 #--------------------------------------------------------------------
```
Comment (gold2718) [inline src/build_scripts/buildlib.csm_share]: These changes do not appear to be related to the PR subject. Please either use a separate PR or document the change in the PR text.

File: config/cesm/machines/Makefile (hunk 1 above)
Comment (jedwards4b) [inline config/cesm/machines/Makefile]: Yes, % symbol in the compset longname is not appreciated by gnu make, but these explicit CURDIR fields aren't needed anyway.

File: src/build_scripts/buildlib.csm_share (hunk 2 above)
Comment (jedwards4b) [inline src/build_scripts/buildlib.csm_share]: PR text updated.

**Comment by gold2718** [comment]
Need citation for new word, "eleminate" #covfefe :)

**Comment by rljacob** [comment]
@jedwards4b, for the full name of an issue id, don't put a space before the # sign.
"Esmci/cime#1718"  not "Esmci/cime #1718"

**Comment by gold2718** [comment]
Suggestion, make [.] illegal in compset names. Above line 546 in case.py, add:
```
        # Period not allowed anywhere in a compset name
        expect("." not in compset,
               "ERROR: period (.) character not allowed in compset name")
```
@mvertens, any reason to not make this change? Checking cesm2_0_alpha07, I don't see any current usage of that character.

File: scripts/tests/scripts_regression_tests.py [hunk 3]
```diff
@@ -1299,6 +1299,12 @@ def test_update_acme_tests_test_mods(self):
         #                  msg="COMMAND SHOULD HAVE WORKED\ncs.status output:\n%s\n\nerrput:\n%s\n\ncode: %d" % (output, errput, stat))
 
 ###############################################################################
+class W_TestCreateTest(TestCreateTestCommon):
+###############################################################################
+    def test_create_test_longname(self):
```
Comment (jgfouca) [inline scripts/tests/scripts_regression_tests.py]: Why not add this to the existing CreateTest test class?

**Comment by jedwards4b** [comment]
Since the testname needs to be parsed before the long compset name is an extra '.' in the compset name in a call to create_test will result in a create_test error: 
```
./create_test SMS.f19_g16.2000_DATM%QI.A_XLND_SICE_SOCN_XROF_XGLC_SWAV
ERROR: Expected 4th item of 'SMS.f19_g16.2000_DATM%QI.A_XLND_SICE_SOCN_XROF_XGLC_SWAV' ('A_XLND_SICE_SOCN_XROF_XGLC_SWAV') to be in form machine_compiler
```
I have added a check in case.py for valid compset names but the create_test code will fail before that point.

File: scripts/tests/scripts_regression_tests.py (hunk 3 above)
Comment (jedwards4b) [inline scripts/tests/scripts_regression_tests.py]: Maybe I'm just being dense - but I don't see one.  Which class specifically?

Comment (jgfouca) [inline scripts/tests/scripts_regression_tests.py]: You're right. The class I was thinking of was TestTestScheduler. That said, this is probably the kind of test I'd put in K_TestCimeCase.

**Comment by gold2718** [comment]
Sigh, 

It does seem like the logic in `parse_test_name` could be tightened up a bit. Can there really be more than 3 dots? Still, if the dot in the compset name is in the right place (e.g., a dot in the GLC component description), it could even pass the test that is in there but be bogus. You could add a check for a legal compset name (maybe add that as a function in utils.py and also call it from case.py) before deciding whether the test name is valid, and if invalid, where the fault probably lies.

BTW, I notice that a common usage of create_test (with a single positional argument) is not in the usage statement. Seems lacking.

**Comment by billsacks** [comment]
Thanks for doing this, @jedwards4b - the ability to use a compset long name in a test will be very helpful! Did you end up using the solution proposed by @gold2718 https://github.com/ESMCI/cime/issues/1718#issuecomment-312312147 of putting the full compset longname in the test name? (That solution makes sense to me.)

**Comment by jedwards4b** [comment]
@billsacks: Yes the full compset longname is in the test name.
@gold2718 I'm not sure what else you want in this PR, could you please clarify.
@jgfouca I've moved the test as suggested, retesting now.

**Comment by gold2718** [comment]
@jedwards4b,

> I'm not sure what else you want in this PR, could you please clarify.

Here is my request (copied from above):

> You could add a check for a legal compset name (maybe add that as a function in utils.py and also call it from case.py) before deciding whether the test name is valid, and if invalid, where the fault probably lies.

To be more specific:
Add a new function, `parse_compset_name`, to utils.py which performs a function similar to `parse_test_name`. `parse_compset_name` will return a list of all the components but also perform error checking (e.g., no dot allowed anywhere in the compset name).
`parse_compset_name` should be called from `parse_test_name` as well as from `get_compset_components` in case.py as well as `_get_description_v3` and `_get_description_match` in component.py. This will ensure that compsets are always parsed and error-checked in a consistent manner.

**Comment by jedwards4b** [comment]
@gold2718 and my reply was that this is a chicken and egg problem.  The code you propose is already there but the parse test name code must be done before the parse compset name code can be called.

**Comment by gold2718** [comment]
@jedwards4b, the reply you reference is not in this conversation after my 
I do not understand why `parse_test_name` cannot call `parse_compset_name` on `rv[2]` before further processing anything in the test name. Also, the rules can be tightened because at least two dots are required (so the expect `num_dots <= 4` should have an accompanying expect of `num_dots > 1`).
While this is not perfect, if `parse_case_name` has sufficient error checking, it will eliminate many of the possible bogus test names which will currently either pass or result in confusing error messages. As a bonus, we have one place where compset names are parsed and checked and that function can have unit test cases built in.

**Comment by jedwards4b** [comment]
@gold2718 I have added tests that may clarify the problem for you, of course future contributions which improve error checking will always be welcome.  This routine does not require that num_dots > 1 as illustrated by the test at line 303 of utils.py

**Comment by gold2718** [comment]
@jedwards4b 
Your first new test case makes my point that it complains about the machine_compiler format when the problem is in fact the compset name. An independent compset name checker would catch most of these errors.
Second, your test on line 303 passes but this is in contradiction to the documentation in both the function (line 300) and in the CIME documentation (http://esmci.github.io/cime/users_guide/testing.html). Either the test case or the documentation is wrong -- please rectify this situation in this PR. The documentation (in both places) requires three elements which implies that two dots are required.

**Comment by gold2718** [comment]
@jedwards4b, if you do not think a common compset name checking function is an improvement (to the current situation of compset names being parsed in three different modules with different error checking), please state your objection here so I can understand.

**Comment by jedwards4b** [comment]
@gold2718 Please explain to me how I would extract and check the compsetname from the testname without first parsing the testname?  

Partial names ('ERS','ERS.f19_f19' are allowed here
and this code is used to parse partial testnames so a check on minimum number of dots would not work.   I have updated the documentation to reflect this.   Note that even if I changed the code to abort on incomplete test names it would not catch the error of an extra dot in the compsetname. 

A more complete check of the compset name is done in create_newcase after the testname has been parsed.

**Comment by gold2718** [comment]
> Please explain to me how I would extract and check the compsetname from the testname without first parsing the testname?

I gave an explicit method above which you adopted with the `check_name` call on line 341.

**Comment by jedwards4b** [comment]
And that works for the second new test, but the first one has an addtional dot and so the compset name is interpreted as 2000_DATM%QI
I believe the resulting error:
   ERROR: Expected 4th item of 'SMS.f19_g16.2000_DATM%QI.A_XLND_SICE_SOCN_XROF_XGLC_SWAV.mach-ine_compiler.test-mods' ('A_XLND_SICE_SOCN_XROF_XGLC_SWAV') to be in form machine_compiler
adequately informs the user as to the problem.

**Comment by gold2718** [comment]
> Partial names ('ERS','ERS.f19_f19' are allowed here
> and this code is used to parse partial testnames so a check on minimum number of dots would not work.

I'm trying to figure out where this happens. One place is the function `_xml_phase`, is this function called anywhere? Why does this function call `parse_test_name` twice for the same info?
Is there any other place where `parse_test_name` is called where a partial test name is okay?

**Comment by gold2718** [comment]
@jedwards4b,

If you had called a compset name checking function on the bogus compset field, `2000_DATM%QI`, it should have kicked an error since compset names currently require more than one underscore. While this is not a perfect check (I think compsets can currently range from 7 to 9 underscores), it would catch the majority of issues and point at the correct problem.

**Comment by gold2718** [comment]
I'm willing to forgo more detailed compset name checking for this PR, please just show me where partial test names are used.

**Comment by jedwards4b** [comment]
@gold2718 A compset name check at this point would also have to allow compset aliases and so a check for the number of underscores wouldn't work.    Partial test names are used in get_full_test_name which is in turn called by the acme test name parsing tools in update_acme_tests.py and scripts_regression_tests.  @jgfouca probably knows more about usage.

**Comment by gold2718** [comment]
Yow, I guess `parse_test_name` is too overloaded for any user-level error checking. How about adding error checking (at least between 2 and 4 dots) after line 286 in create_test?
```
    for test in args.testargs:
        numdots = test.count(".")
        expect(numdots > 1 and numdots <= 4, "Invalid test name, '{}'".format(test))
```
